File: Design-Patterns-Using-Java/Factory/FactoryMethodAnimal.java
import java.util.*;

// Animal interface
interface Animal {
    void speak();
}

// Concrete Animal classes
class Dog implements Animal {
    @Override
    public void speak() {
        System.out.println("Bhow Bhow!!");
    }
}

class Cat implements Animal {
    @Override
    public void speak() {
        System.out.println("Meow Meow!!");
    }
}

// AnimalFactory interface
interface AnimalFactory {
    Animal createAnimal();
}

// Concrete AnimalFactory classes
class CatFactory implements AnimalFactory {
    @Override
    public Animal createAnimal() {
        return new Cat();
    }
}

class DogFactory implements AnimalFactory {
    @Override
    public Animal createAnimal() {
        return new Dog();
    }
}

// Client code
public class FactoryMethodAnimal {
    public static void main(String[] args) {
        AnimalFactory dogFactory = new DogFactory();
        dogFactory.createAnimal().speak();

        AnimalFactory catFactory = new CatFactory();
        catFactory.createAnimal().speak();

        // Future changes to include cow type of objects.
        class Cow implements Animal {
            @Override
            public void speak() {
                System.out.println("Gooaa Gooaa!!");
            }
        }

        class CowFactory implements AnimalFactory {
            @Override
            public Animal createAnimal() {
                return new Cow();
            }
        }

        // Client code for Cow
        AnimalFactory cowFactory = new CowFactory();
        cowFactory.createAnimal().speak();
    }
}
/*
Bhow Bhow!!
Meow Meow!!
Gooaa Gooaa!!
*/

File: Design-Patterns-Using-CSharp/Factory/FactoryMethodAnimal.cs
using System;

// Animal interface
interface IAnimal
{
    void Voice();
}

// Concrete Animal classes
class Dog : IAnimal
{
    public void Voice()
    {
        Console.WriteLine("Bhow Bhow!!");
    }
}

class Cat : IAnimal
{
    public void Voice()
    {
        Console.WriteLine("Meow Meow!!");
    }
}

// AnimalFactory interface
interface IAnimalFactory
{
    IAnimal GetAnimal();
}

// Concrete AnimalFactory classes
class CatFactory : IAnimalFactory
{
    public IAnimal GetAnimal()
    {
        return new Cat();
    }
}

class DogFactory : IAnimalFactory
{
    public IAnimal GetAnimal()
    {
        return new Dog();
    }
}

// Client code
class FactoryMethodAnimal
{
    static void Main(string[] args)
    {
        IAnimalFactory dogFactory = new DogFactory();
        dogFactory.GetAnimal().Voice();

        IAnimalFactory catFactory = new CatFactory();
        catFactory.GetAnimal().Voice();

        // Client code for Cow
        IAnimalFactory cowFactory = new CowFactory();
        cowFactory.GetAnimal().Voice();
    }
}
/*
Bhow Bhow!!
Meow Meow!!
Gooaa Gooaa!!
*/

// Future changes to include cow type of objects.
class Cow : IAnimal
{
    public void Voice()
    {
        Console.WriteLine("Gooaa Gooaa!!");
    }
}

class CowFactory : IAnimalFactory
{
    public IAnimal GetAnimal()
    {
        return new Cow();
    }
}


